File: advection/ftn_code/operator.f90
module operator_mod

    use stvec_mod,  only: stvec_abstract_t
    use params_mod, only: params_t
    use flux_mod,   only: clflux

    implicit none

    type, abstract :: operator_t
        contains
        procedure(act_fun_ifc), deferred :: act_fun
        procedure(act_sub_ifc), deferred :: act_sub
    end type operator_t

    interface
        function act_fun_ifc(this, fin, params) result(fout)
            import operator_t
            import stvec_abstract_t
            import params_t
            class(operator_t),       intent(in)  :: this
            class(stvec_abstract_t), intent(in)  :: fin
            class(params_t),         intent(in)  :: params
            class(stvec_abstract_t), allocatable :: fout
        end function act_fun_ifc

        subroutine act_sub_ifc(this, fout, fin, params)
            import operator_t
            import stvec_abstract_t
            import params_t
            class(operator_t),       intent(in)    :: this
            class(stvec_abstract_t), intent(inout) :: fout
            class(stvec_abstract_t), intent(in)    :: fin
            class(params_t),         intent(in)    :: params
        end subroutine act_sub_ifc
    end interface

    integer(kind=4), parameter :: maxhw = 4

    type, extends(operator_t) :: adv_oper_t
        integer(kind=4)                     :: hw
        procedure(clflux), pointer, nopass  :: fluxfun
        character(:), allocatable           :: flux_scheme_name
        contains
        procedure :: act_fun => adv_oper_fun
        procedure :: act_sub => adv_oper_sub
    end type adv_oper_t

contains

    type(adv_oper_t) function init_adv_operator(namfname) result(op)
        use flux_mod, only: up4f, up1f
        character(*)   :: namfname
        character(256) :: flux_scheme_name="up1"
        namelist /adv/ flux_scheme_name

        open(117,file=namfname, form="formatted")
        read(117,adv)
        close(117)

        if(trim(flux_scheme_name) == "up1") then
            op%fluxfun => up1f
            op%hw = 1
        else if(trim(flux_scheme_name) == "up4") then
            op%fluxfun => up4f
            op%hw = 3
        else
            print *, "Error: unknown flux scheme: "//flux_scheme_name
            stop
        end if

        op%flux_scheme_name = trim(flux_scheme_name)

    end function init_adv_operator

    subroutine adv_oper_sub(this, fout, fin, params)

        use stvec_mod, only: stvec_t
        use flux_mod,  only: up1f,up4f

        class(adv_oper_t),       intent(in)    :: this
        class(stvec_abstract_t), intent(inout) :: fout
        class(stvec_abstract_t), intent(in)    :: fin
        class(params_t),         intent(in)  :: params

        integer(kind=4) N

        select type(fout)
        class is (stvec_t)
        select type (fin)
        class is (stvec_t)
#ifdef ADV_OPT
            if(this%flux_scheme_name == "up1") then
                call flux_conv(fout%p, fin%p, params, up1f, 1)
            else if(this%flux_scheme_name == "up4") then
                call flux_conv(fout%p, fin%p, params, up4f, 3)
            end if
#elif HARD_CODE
            call flux_conv(fout%p, fin%p, params, this%fluxfun, 3)
#else
            call flux_conv(fout%p, fin%p, params, this%fluxfun, this%hw)
#endif
        class default
            print *, "wrong input stvec type in adv_oper_sub"
        end select
        class default
            print *, "wrong out stvec type in adv_oper_sub"
        end select

    end subroutine adv_oper_sub

    function adv_oper_fun(this, fin, params) result(fout)

        use stvec_mod, only: stvec_t
        use flux_mod,  only: up4f

        class(adv_oper_t),       intent(in)  :: this
        class(stvec_abstract_t), intent(in)  :: fin
        class(params_t),         intent(in)  :: params
        class(stvec_abstract_t), allocatable :: fout

        integer(kind=4) :: N

        allocate(stvec_t :: fout)

        select type(fout)
        class is (stvec_t)
        select type (fin)
        class is (stvec_t)
            N = fin%N
            allocate(fout%p(1:N,1:N))
            fout%N = N
            call this%act_sub(fout,fin, params)
        class default
            print *, "wrong input stvec type in adv_oper_fun"
        end select
        class default
            print *, "wrong out stvec type in adv_oper_fun"
        end select
    end function adv_oper_fun

    subroutine flux_conv(fout, fin, params, fluxf, hw)

        class(params_t),        intent(in)    :: params
        real(kind=8),           intent(inout) :: fout(params%N,params%N)
        real(kind=8),           intent(in)    :: fin(params%N,params%N)
        procedure(clflux)                     :: fluxf
        integer(kind=4),        intent(in)    :: hw

        real(kind=8)     :: q(-hw+1:params%N+hw, -hw+1:params%N+hw)
        real(kind=8)     :: flx(0:params%N,params%N)
        real(kind=8)     :: fly(params%N,0:params%N)
        integer(kind=4)  :: N, i, j
#ifdef HARD_CODE
        real(kind=8)     :: za1, za2
#endif

        N = params%N
        call periodic_bc(q,fin,N,hw)

        do j=1,N
            do i=0,N
#ifndef HARD_CODE
                flx(i,j) = fluxf(hw,q(i-hw+1:i+hw,j),params%u(i,j))
#else
                za1 = .5_8+sign(.5_8,params%u(i,j))
                za2 = 1._8-za1
                flx(i,j) = params%u(i,j)*(za1*(3.0_8*q(i+1,j)+13.0_8*q(i  ,j)-5.0_8*q(i-1,j)+q(i-2,j))+ &
                                          za2*(3.0_8*q(i  ,j)+13.0_8*q(i+1,j)-5.0_8*q(i+2,j)+q(i+3,j)))/12.0_8
#endif
            end do
        end do
        do j=0,N
            do i=1,N
#ifndef HARD_CODE
                fly(i,j) = fluxf(hw,q(i,j-hw+1:j+hw),params%v(i,j))
#else
                za1 = .5_8+sign(.5_8,params%v(i,j))
                za2 = 1._8-za1
                fly(i,j) = params%v(i,j)*(za1*(3.0_8*q(i,j+1)+13.0_8*q(i,j  )-5.0_8*q(i,j-1)+q(i,j-2))+ &
                                          za2*(3.0_8*q(i,j  )+13.0_8*q(i,j+1)-5.0_8*q(i,j+2)+q(i,j+3)))/12.0_8
#endif
            end do
        end do

        do j=1,N
            do i=1,N
                fout(i,j) =-(flx(i,j)-flx(i-1,j)+fly(i,j)-fly(i,j-1))/params%dx
            end do
        end do
    end subroutine flux_conv

    subroutine periodic_bc(q,f,N,m)
        real(kind=8),    intent(out) :: q(-m+1:N+m,-m+1:N+m)
        real(kind=8),    intent(in)  :: f(1:N,1:N)
        integer(kind=4), intent(in)  :: N, m

        q(1:N,1:N)     = f(1:N,1:N)
        q(1:N,-m+1:0)  = q(1:N,N-m+1:N)
        q(1:N,N+1:N+m) = q(1:N,1:m)
        q(-m+1:0,-m+1:N+m)  = q(N-m+1:N,-m+1:N+m)
        q(N+1:N+m,-m+1:N+m)  = q(1:m,-m+1:N+m)
    end subroutine periodic_bc

end module operator_mod

File: advection/julia_code/operator.jl
function adv_oper(fin::stvec_t, params::params_t,m::Int,fluxf)
    fout = simil(fin)
    adv_oper!(fout,fin,params,m,fluxf)
    return fout
end
function adv_oper!(fout::stvec_t, fin::stvec_t, params::params_t, m::Int, fluxf)

    #m=1  #simulation of final very optimized code
    q = periodic_bc(fin.q,params.N,m)

    flx = Array{eltype(fin.q)}(undef,params.N+1,params.N)
    fly = Array{eltype(fin.q)}(undef,params.N,params.N+1)

    @inbounds for j=1:params.N
        for i=1:params.N+1
            flx[i,j] = fluxf(m,view(q,i:i+2m-1,j+m),params.u[i,j])
            #hardcoded up4, slightly slower (WTF!?)
            #za1 = 0.5+0.5sign(params.u[i,j])
            #za2 = 1.0-za1
            #flx[i,j] = params.u[i,j]*(za1*(3.0q[i+m,j+m]+13.0q[i+m-1,j+m]-5.0q[i+m-2,j+m]+q[i+m-3,j+m])+
            #                          za2*(3.0q[i+m-1,j+m]+13.0q[i+m,j+m]-5.0q[i+m+1,j+m]+q[i+m+2,j+m]))/12.0

            #flx[i,j] = fluxf(m,q[i:i+2m-1,j+m],params.u[i,j])     #10x slow-down
        end
    end
    @inbounds for j=1:params.N+1
        for i=1:params.N
            fly[i,j] = fluxf(m,view(q,i+m,j:j+2m-1),params.v[i,j])

            #hardcoded up4, slightly slower (WTF!?)
            #za1 = 0.5+0.5sign(params.v[i,j])
            #za2 = 1.0-za1
            #fly[i,j] = params.v[i,j]*(za1*(3.0q[i+m,j+m]+13.0q[i+m,j+m-1]-5.0q[i+m,j+m-2]+q[i+m,j+m-3])+
            #                          za2*(3.0q[i+m,j+m-1]+13.0q[i+m,j+m]-5.0q[i+m,j+m+1]+q[i+m,j+m+2]))/12.0

            #fly[i,j] = fluxf(m,q[i+m,j:j+2m-1],params.v[i,j])     #10x slow-down
        end
    end

    @inbounds for j=1:params.N
        for i=1:params.N
            fout.q[i,j] = -(flx[i+1,j]-flx[i,j]+fly[i,j+1]-fly[i,j])/params.dx
        end
    end
    return
end

@inline function up1(m::Int,q,u)
    za1 = 0.5+0.5sign(u)
    za2 = 1.0-za1
    return @inbounds u*(za1*q[m]+za2*q[m+1])
end
@inline function up4(m::Int,q,u)
    za1 = 0.5+0.5sign(u)
    za2 = 1.0-za1
    return @inbounds u*(za1*(3.0q[m+1]+13.0q[m]-5.0q[m-1]+q[m-2])+
                        za2*(3.0q[m]+13.0q[m+1]-5.0q[m+2]+q[m+3]))/12.0
end
@inline function weno5(m::Int,q,u)
  a1 = 0.5*(1+sign(u))
  a2 = 0.5*(1-sign(u))
  a = a1*q[m-2]+a2*q[m+3]
  b = a1*q[m-1]+a2*q[m+2]
  c = a1*q[m  ]+a2*q[m+1]
  d = a1*q[m+1]+a2*q[m]
  e = a1*q[m+2]+a2*q[m-1]

  q1 = (2.0a-7.0b+11.0c)/6.0
  q2 = (-b+5.0c+2.0d)/6.0
  q3 = (2.0c+5.0d-e)/6.0
  s1 = (13.0/12.0)*(a-2.0b+c)^2 + 0.25(a-4.0b+3.0c)^2
  s2 = (13.0/12.0)*(b-2.0c+d)^2 + 0.25(d-b)^2
  s3 = (13.0/12.0)*(c-2.0d+e)^2 + 0.25(3.0c-4.0d+e)^2
  eps = 1e-6
  w1 = 0.1/(eps+s1)^2
  w2 = 0.6/(eps+s2)^2
  w3 = 0.3/(eps+s3)^2
  return u*(w1*q1+w2*q2+w3*q3)/(w1+w2+w3)
end

function periodic_bc(q,N,m)
    q1 = Array{eltype(q)}(undef,N+2m,N+2m)

    @inbounds q1[m+1:m+N,m+1:m+N] .= q[1:N,1:N]
    @inbounds q1[1:m,m+1:m+N] .= q[N-m+1:N,1:N]
    @inbounds q1[N+m+1:N+2m,m+1:m+N]  .= q[1:m,1:N]
    @inbounds q1[1:N+2m,1:m] .= q1[1:N+2m,N+1:N+m]
    @inbounds q1[1:N+2m,N+m+1:N+2m] .= q1[1:N+2m,m+1:2m]

    return q1
end
